gtp_cmd_handlers.py

```diff
@@ -0,0 +1,340 @@
+    raise ValueError("invalid color {}".format(gtp_color))
+
+
+class BasicCmdHandler(object):
+    """GTP command handler for basic play commands."""
+
+    def __init__(self, player, courtesy_pass=False):
+        self._komi = 6.5
+        self._player = player
+        self._player.initialize_game()
+        self._courtesy_pass = courtesy_pass
+
+    def cmd_boardsize(self, n: int):
+        if n != go.N:
+            raise ValueError("unacceptable size")
+
+    def cmd_clear_board(self):
+        position = self._player.get_position()
+        if (self._player.get_result_string()
+                and position and len(position.recent) > 1):
+            try:
+                sgf = self._player.to_sgf()
+                with open(datetime.datetime.now().strftime("%Y-%m-%d-%H:%M.sgf"), 'w') as f:
+                    f.write(sgf)
+            except NotImplementedError:
+                pass
+            except:
+                print("Error saving sgf", file=sys.stderr, flush=True)
+        self._player.initialize_game(go.Position(komi=self._komi))
+
+    def cmd_komi(self, komi: float):
+        self._komi = komi
+        self._player.get_position().komi = komi
+
+    def cmd_play(self, arg0: str, arg1=None):
+        if arg1 is None:
+            move = arg0
+        else:
+            # let's assume this never happens for now.
+            # self._accomodate_out_of_turn(translate_gtp_color(arg0))
+            move = arg1
+        return self._player.play_move(coords.from_kgs(move))
+
+    def cmd_genmove(self, color=None):
+        if color is not None:
+            self._accomodate_out_of_turn(color)
+
+        if self._courtesy_pass:
+            # If courtesy pass is True and the previous move was a pass, we'll
+            # pass too, regardless of score or our opinion on the game.
+            position = self._player.get_position()
+            if position.recent and position.recent[-1].move is None:
+                return "pass"
+
+        move = self._player.suggest_move(self._player.get_position())
+        if self._player.should_resign():
+            self._player.set_result(-1 * self._player.get_position().to_play,
+                                    was_resign=True)
+            return "resign"
+
+        self._player.play_move(move)
+        if self._player.get_root().is_done():
+            self._player.set_result(self._player.get_position().result(),
+                                    was_resign=False)
+        return coords.to_kgs(move)
+
+    def cmd_undo(self):
+        raise NotImplementedError()
+
+    def cmd_final_score(self):
+        return self._player.get_result_string()
+
+    def _accomodate_out_of_turn(self, color: str):
+        position = self._player.get_position()
+        if translate_gtp_color(color) != position.to_play:
+            position.flip_playerturn(mutate=True)
+
+
+class KgsCmdHandler(object):
+    def __init__(self, player):
+        self._player = player
+
+    def cmd_time_left(self):
```

Comment:
i think this is called with arguments -- its the servers way of informing the engine how much time is left.  There are extended time commands on CGOS that we may want to support for slightly better time management.

Reply:
done.